### ISSUE TITLE
[Question] about nodes

### ISSUE THREAD

**Issue by sitereliabilityengineer** [mingrammer/diagrams#306]
Hello everyone, first of all, apologies if here it's not the principal way to ask my doubts.

I want to dynamically create diagrams. For example, if you have a server that's not healthy, you want to show it with a cross over it or maybe just red around the node.

For example:

`from diagrams.onprem.network import Nginx
from diagrams.onprem.database import PostgreSQL

with Diagram("Advanced Web Service with On-Premise", show=False):
   ingress = Nginx("ingress")

with Cluster("Database HA"):
        master = PostgreSQL("users")
        master - PostgreSQL("slave")  <-- And this server is not healthy`

So "master - PostgreSQL("slave")" is not healthy and I want to show it as non-healthy in the diagram. Which should be the best way to do it ?

**Comment by steschuser** [comment]
Hello,

nice idea.
The easiest way would be to show the connect as broken, I think
That is, change the color from green to red between the systems.

import random
status = random.choice(["red","green"])
master - Edge(labe="connection", color=status) - slave

Or you could try to set the  node_attr of that node
node_attr = = {
    "fontsize": "45",
    "bgcolor": "red"
}
slave - PostgreSQL("slave", node_attr=node_attr) 
https://diagrams.mingrammer.com/docs/guides/diagram
https://www.graphviz.org/doc/info/attrs.html

**Comment by sitereliabilityengineer** [comment]
> Hello,
> 
> nice idea.
> The easiest way would be to show the connect as broken, I think
> That is, change the color from green to red between the systems.
> 
> import random
> status = random.choice(["red","green"])
> master - Edge(labe="connection", color=status) - slave
> 
> Or you could try to set the node_attr of that node
> node_attr = = {
> "fontsize": "45",
> "bgcolor": "red"
> }
> slave - PostgreSQL("slave", node_attr=node_attr)
> https://diagrams.mingrammer.com/docs/guides/diagram
> https://www.graphviz.org/doc/info/attrs.html

Thanks for the help. 

I liked the second option, but it didn't changed the bg color at all:

`
import json

node_attr = {
    "fontsize": "45",
    "bgcolor": "red"
    }

node_attr_json = json.dumps(node_attr)

with Diagram("Advanced Web Service with On-Premise", show=False):
    ingress = Nginx("ingress")

    with Cluster("Database HA"):
        master = PostgreSQL("users")
        master - PostgreSQL("slave", node_attr=node_attr_json)

`
If I don't convert the json to string, when I add the node_attr=node_attr gives an Exception.

After doing the json.dumps no exception is given, but the bg color is the same as default (cluster)

If this is not possible, maybe I should use your first option. 

Thanks.

**Comment by sitereliabilityengineer** [comment]
Hello @steschuser 

Any idea why node_attr  doesn't work for me ?

Thanks.

**Comment by steschuser** [comment]
Hello again,
Im afraid not. I havent actually played around with it yet.

My suggestion would be to check out the genrated .dot file https://github.com/mingrammer/diagrams/pull/182
and check with Graphviz https://github.com/graphp/graphviz

Its on my agenda to take a look into this but time is limited...

**Comment by sitereliabilityengineer** [comment]
Thank you @steschuser !

**Comment by redNixon** [comment]
@sitereliabilityengineer 

The first issue setting bgcolor isn't seem to be the attribute you are really after [the link](https://www.graphviz.org/doc/info/attrs.html) steschuser posted goes into all the differently ways to define a color attribute and when they will matter(honestly I still don't understand its logic fully).  However _how_ exactly to pass in attributes any place besides Diagram() is at best unclear if not confusing to be point of it being called "bugged" in my mind.  Here is a modified version of your code and the resultant diagram:
```python
from diagrams import Cluster, Diagram
from diagrams.onprem.database import PostgreSQL
from diagrams.onprem.network import Nginx

offline_node_attrs = {"fillcolor": "red", "style": "filled"}

with Diagram("Advanced Web Service with On-Premise", show=False) as diag:
    ingress = Nginx("ingress")

    with Cluster("Database HA"):
        master = PostgreSQL("users")

        master - PostgreSQL("slave", **offline_node_attrs)
diag
```
![Screenshot from 2020-10-03 06-14-15](https://user-images.githubusercontent.com/853218/94991112-5f522280-0546-11eb-87cb-a202e20bf363.png)

Besides changing what attributes are being set, notice how the dictionary has to be passed in:
 ```python   
 master - PostgreSQL("slave", **offline_node_attrs)
```
A more standardized name for these attribute variables in the \_\_init__ function of Node, Edge and Cluster would be "kwargs".  To make it clearer about why the dictionary containing attributes needs to be unpacked during the initialization call, consider that this line works the same as the one above and produces the same result:

 ```python   
`master - PostgreSQL("slave", fillcolor="red", style="filled")
````

**Comment by sitereliabilityengineer** [comment]
Hi @redNixon !! That's awesome, thank you so much.

I also have issues with the nodes position within a clúster. I have approx 50 Tomcat nodes in a cluster and when I visualized them, they are all positioned in the same column, so the diagram looks very strange for me.

There's anyway to specify the number of columns in a cluster or change the position of the nodes?

Sorry to ask you all this, but I couldn't find much info.

Thank you again.

**Comment by redNixon** [comment]
Can you post the code?   The answer is yes but what is making it "weird" can often changes and have a simple fix that can be found most often to make graphviz start flowing as expected instead of trying to position everything manually.  When using graphviz indirectly(ie via diagrams) edges, line breaks and transparent clusters are some methods to force some behavior without getting into the graphviz logic though.

I will say that while diagrams is an amazing library to let you quickly code out a diagram, if you start creating real large or have real complex interconnected graphs then utilizing the graphviz library directly will probably be easier(50 nodes is not large btw) to use.

**Comment by sitereliabilityengineer** [comment]
> Can you post the code? The answer is yes but what is making it "weird" can often changes and have a simple fix that can be found most often to make graphviz start flowing as expected instead of trying to position everything manually. When using graphviz indirectly(ie via diagrams) edges, line breaks and transparent clusters are some methods to force some behavior without getting into the graphviz logic though.
> 
> I will say that while diagrams is an amazing library to let you quickly code out a diagram, if you start creating real large or have real complex interconnected graphs then utilizing the graphviz library directly will probably be easier(50 nodes is not large btw) to use.

Hey @redNixon 

Sure, my code is not complex.
`
from diagrams import Cluster, Diagram
from diagrams.aws.network import ELB
from diagrams.aws.network import Route53
from diagrams.onprem.network import Tomcat

graph_attr = {
    "fontsize": "25",
    "bgcolor": "red"
}

with Diagram("DC1 DIAGRAM", show=False):
    dns = Route53("dns")
    lb = ELB("lb")


    with Cluster("Application servers"):
        app_group = [
                Tomcat("server-app1"),
                Tomcat("server-app2"),
                Tomcat("server-app3"),
                ......
                Tomcat("server-app50")
                ]

        with Cluster("IN_MAINTENANCE", graph_attr=graph_attr):
            app_maintenance = [Tomcat("server-app14)]

    with Cluster("DB Cluster"):
        db_master = RDS("userdb")
        db_master - [RDS("userdb ro")]

    memcached = ElastiCache("memcached")

    dns >> lb >> app_group
    app_group >> db_master
    app_group >> memcached

`

So basically, app_group appear all in the same column, with a few nodes it's ok, but when you have 50 nodes in the same column, it's not great.

Thank you!

**Comment by steschuser** [comment]
This looks like the problem in https://github.com/mingrammer/diagrams/issues/276

**Comment by sitereliabilityengineer** [comment]
Hello @steschuser 

Thanks a lot for your help!!

**Comment by sitereliabilityengineer** [comment]
Tried adding size & ratio, but I can't see any difference.

**Comment by clayms** [comment]
> I also have issues with the nodes position within a clúster. I have approx 50 Tomcat nodes in a cluster and when I visualized them, they are all positioned in the same column, so the diagram looks very strange for me.
> 
> There's anyway to specify the number of columns in a cluster or change the position of the nodes?


@sitereliabilityengineer 

See if you can adapt the following to fit your needs:

```
from diagrams import Diagram, Cluster, Edge, Node

graph_attr = {
    "layout":"neato",
    }

clus_attr = {
    "bgcolor":"transparent",
    "pencolor":"blue",
    "penwidth":"4.0"
    }

with Diagram("Clusters", show=False, graph_attr=graph_attr) as diag:

    with Cluster("A"):
        A_UpLf = Node("", shape="plaintext", pin="true", pos="0,4")
        A_LwRt = Node("", shape="plaintext", pin="true", pos="4,0")
        A1 = Node("A1", shape="circle", labelloc="c", pin="true", pos="1,0")
        A2 = Node("A2", shape="triangle",    labelloc="c", pin="true", pos="3,3.5")
    with Cluster("B"):
        B_UpLf = Node("", shape="plaintext", pin="true", pos="6,4")
        B_LwRt = Node("", shape="plaintext", pin="true", pos="10,-0.5")
        B1 = Node("B1", shape="star", labelloc="c", pin="true", pos="9,4")
        B2 = Node("B2", shape="box3d",    labelloc="c", pin="true", pos="7,1.5")
    with Cluster("C", graph_attr=clus_attr):
        C_UpLf = Node("", shape="plaintext", pin="true", pos="2,3")
        C_LwRt = Node("", shape="plaintext", pin="true", pos="8,-2")
        C1 = Node("C1", shape="folder", labelloc="c", pin="true", pos="6,2.5")
        C2 = Node("C2", shape="rarrow", labelloc="c", pin="true", pos="2.5,-1")

diag
```


![image](https://user-images.githubusercontent.com/12793783/98180722-b47bae00-1ecf-11eb-8d27-f4e7ef6bc532.png)
